File: sdk/ios/Sources/FrakSDK/Identity/IdentityMerge.swift
import CryptoKit
import Foundation

/// Inbound `?fmt=` handling: this install is the merge *target*, folding its anonymous id into the
/// group the token names. Owns token parsing, the wire-body shape and the same-process claim; the
/// post itself runs through `MergeSender` off the durable queue.
actor IdentityMerge {
    static let tokenKey = "fmt"
    static let executePath = "/user/identity/merge/execute"

    private let logger: FrakLogger
    private var consumed: Set<String> = []
    private var arrivals: Set<String> = []

    init(logger: FrakLogger) {
        self.logger = logger
    }

    static func parseToken(_ url: String) -> String? {
        URLQuery.parse(url)?.value(for: tokenKey).flatMap { $0.isEmpty ? nil : $0 }
    }

    /// The bytes a merge proof signs over. Shared with `MergeSender` so the queued wire post and
    /// the enclave binding never drift on what is being attested to.
    static func binding(_ mergeToken: String) -> Data {
        Data(SHA256.hash(data: Data(mergeToken.utf8)))
    }

    /// The wire body for `executePath`. Nil only if `JSONSerialization` itself fails, which none
    /// of these plain-string fields can trigger in practice.
    static func body(mergeToken: String, anonymousId: String, merchantId: String, proof: String) -> Data? {
        let payload: [String: String] = [
            "mergeToken": mergeToken,
            "targetAnonymousId": anonymousId,
            "merchantId": merchantId,
            "proof": proof,
        ]
        return try? JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys])
    }

    /// Burns a merge token once. In-memory only — `EventOutbox.isQueued` survives a restart, this
    /// closes the race two same-process `handleReferralLink` calls win against a disk check.
    func claim(_ mergeToken: String) -> Bool {
        !mergeToken.isEmpty && consumed.insert(mergeToken).inserted
    }

    /// The arrival half of `claim`: SwiftUI delivers `.onOpenURL` to every view registering the
    /// modifier, so one tap reaches here many times and each would track its own arrival.
    func claimArrival(_ context: String) -> Bool {
        !context.isEmpty && arrivals.insert(context).inserted
    }
}

File: sdk/android/frak-sdk/src/main/kotlin/id/frak/sdk/identity/IdentityMerge.kt
package id.frak.sdk.identity

import id.frak.sdk.net.UrlQuery
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.json.JSONObject
import java.security.MessageDigest

/**
 * Inbound `?fmt=` wire contract: this install is the merge *target*, folding its anonymous id into
 * the group the token names. Keeps only the once-per-process claim and the stateless wire
 * contract; delivery lives in [id.frak.sdk.tracking.MergeSender].
 */
internal class IdentityMerge {
    private val mutex = Mutex()
    private val consumed = mutableSetOf<String>()
    private val arrivals = mutableSetOf<String>()

    /**
     * Claims a token for this process, so a router replaying the same intent on every activity
     * recreation enqueues one merge. Separate from delivery, which the durable queue retries.
     */
    suspend fun claim(mergeToken: String): Boolean =
        mergeToken.isNotEmpty() && mutex.withLock { consumed.add(mergeToken) }

    /**
     * The arrival half of [claim]: a merchant routing the same URL by hand from more than one
     * entry point would otherwise track one arrival per call.
     */
    suspend fun claimArrival(context: String): Boolean =
        context.isNotEmpty() && mutex.withLock { arrivals.add(context) }

    companion object {
        const val TOKEN_KEY: String = "fmt"
        const val MERGE_EXECUTE_PATH: String = "/user/identity/merge/execute"

        fun parseToken(url: String): String? = UrlQuery.parse(url)?.getExact(TOKEN_KEY)?.takeIf { it.isNotEmpty() }

        /** UTF-8, matching `IdentityProofService.hashMergeToken`. Binds the proof to this token, not just this merchant. */
        fun binding(mergeToken: String): ByteArray =
            MessageDigest.getInstance("SHA-256").digest(mergeToken.toByteArray(Charsets.UTF_8))

        fun body(
            mergeToken: String,
            anonymousId: String,
            merchantId: String,
            proof: String,
        ): JSONObject =
            JSONObject()
                .put("mergeToken", mergeToken)
                .put("targetAnonymousId", anonymousId)
                .put("merchantId", merchantId)
                .put("proof", proof)
    }
}
